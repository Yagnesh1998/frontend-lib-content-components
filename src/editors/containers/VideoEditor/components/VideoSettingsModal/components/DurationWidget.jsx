import React from 'react';
import { useDispatch } from 'react-redux';

import {
  Col,
  Form,
  Row,
} from '@edx/paragon';

import { injectIntl, intlShape, FormattedMessage } from '@edx/frontend-platform/i18n';
import { keyStore } from '../../../../../utils';
import CollapsibleFormWidget from './CollapsibleFormWidget';
import hooks from './hooks';
import { durationFromValue } from './duration';
import messages from './messages';

/**
 * Collapsible Form widget controlling video start and end times
 * Also displays the total run time of the video.
 */
export const DurationWidget = ({
  // injected
  intl,
}) => {
  const dispatch = useDispatch();
  const { duration } = hooks.widgetValues({
    dispatch,
    fields: { [hooks.selectorKeys.duration]: hooks.durationWidget },
  });
  const timeKeys = keyStore(duration.formValue);

  const getTotalLabel = (startTime, stopTime, subtitle) => {
    if (!stopTime) {
      if (!startTime) {
        return intl.formatMessage(messages.fullVideoLength);
      }
      if (subtitle) {
        return intl.formatMessage(messages.startsAt, { startTime: durationFromValue(startTime) });
      }
      return null;
    }
    const total = stopTime - (startTime || 0);
    return intl.formatMessage(messages.total, { total: durationFromValue(total) });
  };

  return (
    <CollapsibleFormWidget
      title={intl.formatMessage(messages.durationTitle)}
      subtitle={getTotalLabel(duration.formValue.startTime, duration.formValue.stopTime, true)}
    >
      <FormattedMessage {...messages.durationDescription} />
      <Row className="mt-4">
        <Form.Group as={Col}>
          <Form.Control
            floatingLabel={intl.formatMessage(messages.startTimeLabel)}
            value={duration.local.startTime}
            onBlur={duration.onBlur(timeKeys.startTime)}
            onChange={duration.onChange(timeKeys.startTime)}
          />
          <Form.Control.Feedback>
            <FormattedMessage {...messages.durationHint} />
          </Form.Control.Feedback>
        </Form.Group>
        <Form.Group as={Col}>
          <Form.Control
            floatingLabel={intl.formatMessage(messages.stopTimeLabel)}
            value={duration.local.stopTime}
            onBlur={duration.onBlur(timeKeys.stopTime)}
            onChange={duration.onChange(timeKeys.stopTime)}
          />
          <Form.Control.Feedback>
            <FormattedMessage {...messages.durationHint} />
          </Form.Control.Feedback>
        </Form.Group>
      </Row>
      <div className="mt-4">
        {getTotalLabel(duration.formValue.startTime, duration.formValue.stopTime)}
      </div>
    </CollapsibleFormWidget>
  );
};

DurationWidget.propTypes = {
  // injected
  intl: intlShape.isRequired,
};

export default injectIntl(DurationWidget);
