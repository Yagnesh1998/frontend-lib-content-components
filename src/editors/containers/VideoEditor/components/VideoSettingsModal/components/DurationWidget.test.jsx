import React from 'react';
import { shallow } from 'enzyme';

import { formatMessage } from '../../../../../../testUtils';
import { DurationWidget } from './DurationWidget';

describe('DurationWidget', () => {
  const props = {
    isError: false,
    subtitle: 'SuBTItle',
    title: 'tiTLE',
    // inject
    intl: { formatMessage },
  };
  describe('render', () => {
    const testContent = (<p>Some test string</p>);
    test('snapshots: renders as expected with default props', () => {
      expect(
        shallow(<DurationWidget {...props} />),
      ).toMatchSnapshot();
    });
  });
});
