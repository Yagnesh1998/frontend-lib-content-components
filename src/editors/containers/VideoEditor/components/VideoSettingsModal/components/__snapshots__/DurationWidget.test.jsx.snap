// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DurationWidget render snapshots: renders as expected with default props 1`] = `
<injectIntl(ShimmedIntlComponent)
  subtitle="Full video length"
  title="Duration"
>
  <FormattedMessage
    defaultMessage="Set a specific section of the video to play."
    description="Description of Duration widget"
    id="authoring.videoeditor.duration.description"
  />
  <Component
    className="mt-4"
  >
    <Form.Group>
      <Form.Control
        floatingLabel="Start time"
        onBlur={[Function]}
        onChange={[Function]}
        value="00:00:00"
      />
      <Form.Control.Feedback>
        <FormattedMessage
          defaultMessage="Enter time as HH:MM:SS"
          description="Hint text for start and stop time input fields"
          id="authoring.videoeditor.duration.hint"
        />
      </Form.Control.Feedback>
    </Form.Group>
    <Form.Group>
      <Form.Control
        floatingLabel="Stop time"
        onBlur={[Function]}
        onChange={[Function]}
        value="00:00:00"
      />
      <Form.Control.Feedback>
        <FormattedMessage
          defaultMessage="Enter time as HH:MM:SS"
          description="Hint text for start and stop time input fields"
          id="authoring.videoeditor.duration.hint"
        />
      </Form.Control.Feedback>
    </Form.Group>
  </Component>
  <div
    className="mt-4"
  >
    Full video length
  </div>
</injectIntl(ShimmedIntlComponent)>
`;
